WalletWasabi/WabiSabi/Crypto/ZeroCredentialPool.cs

```diff
@@ -0,0 +1,58 @@
+using System.Linq;
+using System.Collections.Generic;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using System.Collections.Concurrent;
+
+namespace WalletWasabi.WabiSabi.Crypto
+{
+	/// <summary>
+	/// Keeps a registry of credentials available.
+	/// </summary>
+	public class ZeroCredentialPool
+	{
+		private readonly ConcurrentQueue<Credential> _zeroValueCredentials = new ();
+
+		internal ZeroCredentialPool()
+		{
+		}
+
+		/// <summary>
+		/// Registers those zero value credentials that were issued and return the valuable credentials.
+		/// </summary>
+		/// <param name="credentials">Credentials received from the coordinator.</param>
+		public IEnumerable<Credential> RegisterZeroValueCredentials(IEnumerable<Credential> credentials)
+		{
+			foreach (var credential in credentials)
+			{
+				if (credential.Amount.IsZero)
+				{
+					_zeroValueCredentials.Enqueue(credential);
+				}
+				else
+				{
+					yield return credential;
+				}
+			}
+		}
+
+		public Credential[] CompleteWithZeroCredentials(IEnumerable<Credential> credentials)
+		{
+			var n = ProtocolConstants.CredentialNumber;
+			var credentialsToReturn = new Credential[ProtocolConstants.CredentialNumber];
+			foreach (var credential in credentials.Take(ProtocolConstants.CredentialNumber).ToArray())
+			{
+				n--;
+				credentialsToReturn[n] = credential;
+			}
+			for (; n > 0; )
+			{
+				if (_zeroValueCredentials.TryDequeue(out var credential))
```

Comment:
Can be a deadlock. Add some conditions to throw sooner or later.

Reply:
Can be a cancellationToken.

Reply:
Or you can throw if we are running out of ZeroCreds but it is hard to decide that it was because of concurrency or because of a lack of zero creds. 

`If two threads call TryDequeue at precisely the same moment, neither operation is blocked. When a conflict is detected between two threads, one thread has to try again to retrieve the next element, and the synchronization is handled internally.`
https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1.trydequeue?view=net-5.0#remarks

Reply:
when reissuance requests always give `k * (k-1)` zero credentials the only reason to not get them is a concurrency issue

Reply:
**Clarification:** the PR was developed under the assumption that we can never run out of null credentials. If for some reason we run out of them then we have a bug somewhere.

I agree with @molnard that we have to prevent entering in an infinite loop. We can use a cancellation token or max retry attempts. Let me check which one is less intrusive.

Reply:
I think that assumption is correct, the whole point of decreeing that K is a constant was to legitimize this kind of simplification.

We could wrap the queue with some domain specific abstraction that enforces this, where the `Dequeue` operation can't fail, and throws an exception if the number of pending credentials is ever lower than the number of dequeue operations (so we can immediately add a `Task<Credential>` (maybe a `TaskCompletionSource`) which gives a zero credential to the pool as soon as it's available. Or maybe there's a simpler way of enforcing this invariant.

Reply:
Ok. I've modified this to support up to 100 retries. Why? because:

* we should never run out of null credentials. If this happens then we have a bug in our implementation and we have to fix the real cause.
* `ConcurrentQueue` handles the concurrency correctly and even when no thread is blocked in `TryDequeue` that doesn't mean that them can take the same credential. One takes it and the other spins an retry (I have taken a look at the source code).
* `CancellationToken` is more elegant but this is a extremely-short-live operation that should never fail and passing a CancellationToken doesn't seem to worth it.

Is this okay for you?

Reply:
Why not use a bounded channel instead of a concurrent queue, with a bound `dependencyGraph.Vertices.Length * k`? That's guaranteed to never block for the writer, and won't need any sort of retry logic for the readers. At the very least the queue could be paired with a cond var so that enqueueing wakes up the listeners who then do this spin loop several times or go back to waiting on the cond var if they lost the race to dequeue.

100 seems arbitrary and might not be enough if the task to enqueue is not scheduled for some reason (e.g. a some interaction with random delays). I don't like these sorts of temp hacks in any sort of concurrent code as in my experience they always lead to regret eventually, even when some future bug only might be related to that but actually turns out to be elsewhere, debugging and convincing yourself that such a thing is really not the the cause of problem it only might be related to is harder than just doing it properly in the first place, except it inevitably happens when the timing is worse.

This code might end up interacting with the random delays, or other concurrency aspects of the round client or the wallet as a whole. Using a blocking abstraction for dequeue is almost the same code but eliminates the need for a spin loop, and means we won't have to worry about triggering errors based on how the round client concurrency is done (i.e. when exactly the Alices and Bobs are constructed and their methods used, and how delays are calculated and accounted for).

Also bear in mind that we will need to send cover traffic of reissuances with spare zero credentials, and even when a client doesn't need to reissue, sending unneeded ones increases their anonymity sets depending on the input amounts, because in a scenario where several much smaller inputs or one much larger input was broken down to produce the user's desired output registrations, that hypothetical user would have had to do reissuance, so could not have made those output registrations before reissuing. Each such unnecessary reissuance adds K additional spare credentials (and when the channel is full they can just be dropped, but that shouldn't happen) but initially they also need zero credentials which in the case of a fan out decomposition might cause contention because reissuances happen concurrently with output registrations in general.

Reply:
I didn't use a channel because they are designed for situations when the producer and the consumer are independent and the consumer can need to wait for the producer to produce or the producer could need to stop producing because the workload it too much for the consumer. This was not the assumption for my design, I always assumed that we have the null credentials already available and the retry is just to prevent entering in an infinite loop because of some kind of bug.

What you suggest is to work with a different assumption and redesign the pool with the idea that nulls credentials could be not available and then the caller would need to wait for them, what requires doing this async.

Reply:
No, i think this assumption still holds, it's just that it meaningfully translates to an expectation that the only cause for any blocking is the result of thread scheduling, whereas with the concurrent queue that's not taken into account even though it's inherently there.

Debugging race conditions is really a lot of work so avoiding it by writing things to be be safe from a concurrency point of view is always desirable IMO, but because of the assumption of sufficient credentials we can be assured that it should almost never block in practice, and if/when it does it will it will be for a very short time and never because of anything initiated by the round client or its sub-objects.

Not writing to tolerate failure way means we're also assuming there's no unforeseen performance issues that originate from the platform, the OS, and the hardware, it's a huge surface for generating interference. The consequences of an occasional swap thrash or GC hiccup might mean clients just error and fail to follow through on the round so if users ever suffer from it it's also very low that we will find out and manage to pin the cause, because even replicating the failure is hard and relies on non-deterministic circumstances which are by design low probability, hard to tell apart from noise.

The assumption of always having spare credentials still ensures that we don't need to address a higher level problem, of generating a deadlock free depdency graph with the additional constraint that zero credentials are always eventually available, which is still a huge simplification and reduction in scope.

Reply:
Just to be clear, my opposition is only to this fix.

I think we're better off with leaving the possibility of an infinite loop bug, and personally i'm happy enough with the high level reasoning for why the infinite loop should never happen in practice, but if we want to address it then that should be done without exchanging it for a potential concurrency bugs for which the equivalent argument isn't that it should never happen, but that it should only happen under very strange circumstances.

Reply:
okay, I don't get it. Take a look at the latest commit 77596d60c5ee2e69f1a8ca03d2a2e5acb31e350c and let me know if it has to be reverted or not.

Reply:
If there are always enough Zero creds then there is no need to wait - if there are none, just throw. I would solve the concurrency issue by using lock instead of concurrent. So the implementation would be very similar.

Reply:
There always will be enough zero credentials in principle, but we can't guarantee that at any moment there's already ones in the queue without synchronizing the different tasks, because threadpool scheduling is non deterministic.

If a thread needs a zero credential it can be sure that if it begins dequeuing then eventually it will succeed, but without synchronization we can't be sure when it will succeed, so an infinite loop is correct. We can possibly with some cancellation condition, but throwing an exception or stopping that loop at an arbitrary point is simply not correct because it can non-deterministically fail.

Because of the assumption we can ignore the inefficiency of the spin loop, as this should not happen often, but if we want to do using synchronization primitives, the correct abstraction is a condition variable, which if I understand correctly in C# is the `Monitor` class (specifically `Pulse`).

Reply:
I will do what @molnard suggests because it is the simplest thing to do and should be enough. In case our assumptions are wrong then we can implement something different. I will also use a lock, even when it is not necessary, just to unblock all this.

Reply:
Hmm, I tried to reply by email but it's not going through.

Nack from me re lock, it doesn't actually fix the issue, the concurrent queue is already thread safe the problem is what happens if it's empty which we can't guarantee won't happen, we can only guarantee that if it's empty it won't stay empty for long.

I'm happy to ack:

1. just leave the infinite loop in there, it's more correct than having a spin loop
2. use a semaphore, increment when enqueining and decrement when dequeueing, and a reader should only dequeue after successfully decrementing the semaphore.
3. use some other synchronization primitive, such as a cond var, some sort of channel based construction, or whatever (thread blocking or async, doesn't matter).

But it's just not reasonable to mitigate a hypothetical bug for which we have no evidence that it's a reasonable outcome but we have very good reasons to claim it shouldn't ever be an issue by introducing clearly incorrect but confusing behaviour which for sure can happen, just rarely.

If the infinite loop thing is a real concern (and I'm really not convinced that it is) then we should not solve this by some weird mishmash of vaguely related primitives, this is a thread synchronization problem and it should be solved with the appropriate primitives.

Reply:
I don't know what to do now. All this feels like playing rock/paper/scissors where whatever you chose there is better hand and in the end there is not way to win. 

I think the first solution (the loop) was enough. I agree with @molnard that it would be convenient to prevent infinite loops just in case, even when we think it is not possible so, the second solution (loop with a circuit breaker) was okay too or even better.

Reply:
the loop on its own is OK because it correctly waits an unbounded amount of time for an event which we designed to protocol to ensure should happen during normal operation. we can't control ephemeral network errors, thread scheduling, or a number of other things, but when protocol messages are delivered in a timely manner it's not possible to run out of zero credentials and become deadlocked.

if we ever run out of zero credentials this is because there is a request that is currently in flight or which should be retried shortly. this can and will happen unless we ensure to only create clients after we already have the credentials, in which case this really shouldn't go through a queue but explicitly schedule which zero credential goes to what client and when.

in the event that such a failure happens with the circuit breaker, since it's a spin loop it's very likely that the 100 (or really arbitrarily many) iterations of the loop can happen before some thread manages to enqueue another. if we let the loop run unbounded we're wasting CPU as that happens, but that's all we're wasting and it should happen rarely enough that it doesn't matter.

but if we're breaking out of the loop, then on the rare occasion that it does happen, this "fix" makes it into an unrecoverable error. that's a much more serious bug than a rare hang (which we have every reason to believe won't happen), because it **WILL** happen eventually, and it will cost us a lot more time than we've already wasted discussing this.

if you really insist on fixing the infinite loop (i still think that's a mistake, the loop is 100% correct IMO) then the do that is synchronize the threads so that the data dependency is reflected in the execution order, and it's not just a question of exclusive access to the queue, it's that the readers of the queue have to block indefinitely until there are items available to dequeue. again, this is trivial to achieve with a semaphore, and it can't be done with just locks, it needs some kind of waiting primitive. the last commit just reimplements the same semantics as the concurrent queue.

Reply:
From my maintenance experience, I learned that the infinite loop is one of the worst errors ever. You don't know what is happening but the service is not responding. I also learned that no matter how small the possibility if there is no prevention it will happen.

Reply:
There is prevention of an infinite loop, it's just indirect, the condition is that when an item is added to the queue and the current thread manages to dequeue it, the loop stops.

For me that is good enough, but I can see the arguments for why maybe that's not obvious enough, etc. However, in that case we need to fix it using a semaphore or some sort of blocking queue, etc, but breaking after n iterations is **A BUG**, plain and simple.

I really can't believe we're still having this conversation =(

Reply:
you're right, although that's like setting the max iters to 1 ;-)